src/reduce.jl

```diff
@@ -22,24 +22,32 @@ reduce_op(op::AddSubMul) = add_sub_op(op)
 
 reduce_op(::typeof(add_dot)) = +
 
-neutral_element(::typeof(+), T::Type) = zero(T)
-
 map_op(::AddSubMul) = *
 
 map_op(::typeof(add_dot)) = LinearAlgebra.dot
 
-function promote_map_reduce(op::Function, args::Vararg{Any,N}) where {N}
-    return promote_operation(
-        op,
-        promote_operation(map_op(op), args...),
-        args...,
-    )
+# We need a generated function here to help with type stability.
+@generated function _accumulator(f::F, op::O, args::Vararg{Any,N}) where {F,O,N}
+    expr = Expr(:call, :(map_op(op)))
+    for i in 1:N
+        push!(expr.args, :(zero(f(args[$i]))))
+    end
+    return expr
 end
 
 function fused_map_reduce(op::F, args::Vararg{Any,N}) where {F<:Function,N}
     _check_same_length(args...)
-    T = promote_map_reduce(op, eltype.(args)...)
```

Comment:
What about `T = promote_map_reduce(op, _concrete_eltype.(args)...)` where `_concrete_eltype` checks for the first elements etc... ? At least this gives a function that we can easily reuse in the rest of MA to have something consistent.
Also, we then just have to add a `promote_operation` method for arrays

Reply:
Updated to use your suggestion.